modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/database/inlinecolumn/InlineIndexColumnFactory.java

```diff
@@ -0,0 +1,236 @@
+        Value.FLOAT,
+        Value.INT,
+        Value.LONG,
+        Value.JAVA_OBJECT,
+        Value.SHORT,
+        Value.STRING,
+        Value.STRING_FIXED,
+        Value.STRING_IGNORECASE,
+        Value.TIME,
+        Value.TIMESTAMP,
+        Value.UUID
+    ));
+
+    /** */
+    private final ConcurrentHashMap<String, InlineIndexColumn> helpers = new ConcurrentHashMap<>();
+
+    /**
+     * Creates inline column wrapper for given indexed column.
+     *
+     * @param col Column.
+     */
+    private static InlineIndexColumn createHelper(CompareMode mode, Column col) {
+        boolean useOptimizedStrComp = CompareMode.OFF.equals(mode.getName());
+
+        switch (col.getType().getValueType()) {
+            case Value.BOOLEAN:
+                return new BooleanInlineIndexColumn(col);
+
+            case Value.BYTE:
+                return new ByteInlineIndexColumn(col);
+
+            case Value.SHORT:
+                return new ShortInlineIndexColumn(col);
+
+            case Value.INT:
+                return new IntegerInlineIndexColumn(col);
+
+            case Value.LONG:
+                return new LongInlineIndexColumn(col);
+
+            case Value.FLOAT:
+                return new FloatInlineIndexColumn(col);
+
+            case Value.DOUBLE:
+                return new DoubleInlineIndexColumn(col);
+
+            case Value.DATE:
+                return new DateInlineIndexColumn(col);
+
+            case Value.TIME:
+                return new TimeInlineIndexColumn(col);
+
+            case Value.TIMESTAMP:
+                return new TimestampInlineIndexColumn(col);
+
+            case Value.UUID:
+                return new UuidInlineIndexColumn(col);
+
+            case Value.STRING:
+                return new StringInlineIndexColumn(col, useOptimizedStrComp);
+
+            case Value.STRING_FIXED:
+                return new FixedStringInlineIndexColumn(col, useOptimizedStrComp);
+
+            case Value.STRING_IGNORECASE:
+                return new StringIgnoreCaseInlineIndexColumn(col, useOptimizedStrComp);
+
+            case Value.BYTES:
+                return new BytesInlineIndexColumn(col, mode.isBinaryUnsigned());
+
+            case Value.JAVA_OBJECT:
+                return new ObjectHashInlineIndexColumn(col);
+        }
+
+        throw new IllegalStateException("Unknown value type=" + col.getType().getValueType());
+    }
+
+    /** */
+    private final CompareMode mode;
+
+    /**
+     * @param col Column.
+     * @param inlineObjHash Inline object hash.
+     */
+    public InlineIndexColumn createInlineHelper(Column col, boolean inlineObjHash) {
+        int type = col.getType().getValueType();
+
+        if (type == Value.JAVA_OBJECT && !inlineObjHash)
+            // fallback to old inlining logic
+            return new ObjectBytesInlineIndexColumn(col, mode.isBinaryUnsigned());
+
+        return helpers.computeIfAbsent(col.getName(), name -> createHelper(mode, col));
```

Comment:
Is there any place where creating inlineHelper could be concurrent? I haven't found such place.

Reply:
no, it currently used in synchronous way. I'll change type of the map to HashMap so as not to be misleading